Add a style guide section (#73)
exasim-project/NeoFOAM: doc/contributing.rst

Contributions are highly welcomed. Here is some information getting you started.

NeoFOAM Code Style Guide
""""""""""""""""""""""""

 * use `camelCase` for functions and members and capitalized `CamelCase` for classes

>  * use `camelCase` for files and folder names

Author: yes.

>  * use `camelCase` for files and folder names

Author: So basically everything starts with a lower case. Except class names, abbreviations like CPU, GPU and NeoFOAM.hpp

Contributions are highly welcomed. Here is some information getting you started.

NeoFOAM Code Style Guide
""""""""""""""""""""""""

To simplify coding and code reviews you should adhere to the following code style. However, most
of NeoFOAMs code style guide is checked via automated tools.
For example, formatting and non-format related code style is enforced via clang-format and clang-tidy.
Corresponding configuration files are `.clang-format <https://github.com/exasim-project/NeoFOAM/blob/main/.clang-format>`_
and `.clang-tidy <https://github.com/exasim-project/NeoFOAM/blob/main/.clang-format/.clang-tidy>`_.
Furthermore, adequate licensing of the source code is checked via the reuse linter and typos checks for obvious spelling issues.
Thus, this style guide doesn't list all stylistic rules explicitly but rather gives advise for ambiguous situations and mentions the rational for some decissions.

>  * Use `camelCase` for functions and members and capitalized `CamelCase` for classes

Reviewer: For members are we also post fixing '_'? I have been doing it because I think I saw Henning do it - perhaps we could just clarify that here.

 * Use descriptive template parameter names.
   For example prefer `template <class ValueType>` over `template <class T>` if the type can be a float or double.
 * Use `[[nodiscard]]` except for getters.
   To indicate that a function simply returns a value instead of performing expensive computations `[[nodiscard]]` can be omitted.
 * Use `final` except for abstract classes.
   NeoFOAMs aims at having a rather flat inheritance hierarchy and composition is generally preferred.
   To avoid unintended inheritance the using `final` is advised.
 * Avoid storing references as data members in classes.
   An exception might be a `const Unstructured& mesh` as it will most likely outlive any other objects.

>  * Place in-out parameter at the end of function argument lists.
>    E.g. `foo(const Type& in, Type& out)` instead of `foo(Type& out,const Type& in)`.

Reviewer: ```suggestion
 * Order of parameters to functions should be in, in-out, out.
   E.g. `foo(const Type& in, Type& inOut, Type& out)` instead of `foo(Type& inOut,  Type& out, const Type& in)` or some other variation.
```
